### PR TITLE
feat: add product selector mode

### PR DESCRIPTION
**Pull request by rs-amp**
Adds a product selector to the extension. Supports providing an array of products or a single one, in either string format (supporting pattern) or as object containing product and variant ID.

Supports searching via category or query.

### CODE REVIEW

File: components/ProductSelector/index.tsx [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+		setResults([])
+	}, [mode])
+
+	useEffect(() => {
+		setPage(1)
+		setNoOfPages(Math.ceil(results.length / itemsPerPage))
+		setTimeout(() => {
+			getContainerHeight()
+		}, 100)
+		//console.log('results: ', results)
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:215]: unused?

File: components/ProductSelector/index.tsx [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+	}, [results, selectedProducts])
+
+	useEffect(() => {
+		window.addEventListener('resize', getContainerHeight)
+
+		return () => {
+			window.removeEventListener('resize', getContainerHeight)
+		}
+	}, [])
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:224]: getContainerHeight is not captured here. Maybe that should also be a useCallback dependant on `container`.

File: components/ProductSelector/index.tsx [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+					)
+					return {
+						...item,
+						deleteKey: index,
+						selectedVariant:
+							selVar !== undefined ? selVar : item.variants[0],
+					}
+				})
+
+				//if(ampSDK?.type === 'object') console.log('loaded object:', prod)
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:281]: unused?

File: components/ProductSelector/index.tsx [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+			<TabPanel value={mode} index={0}>
+				<Box sx={{ display: 'flex' }}>
+					<TextField
+						inputRef={keywordInput}
+						size='small'
+						sx={{ ml: 1, flex: 1 }}
+						label='Keyword Search (esc. to clear)'
+						variant='outlined'
+						inputProps={{
+							'aria-label': 'keyword search (escape key to clear',
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:348]: Missing `)`

File: components/ProductSelector/index.tsx [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+			) : (
+				<></>
+			)}
+
+			{/* Dual Mode Search */}
+
+			<Tabs
+				value={mode}
+				onChange={tabChange}
+				aria-label='basic tabs example'
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:333]: This seems like a placeholder.

File: components/ProductSelector/index.tsx [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+						width: '100%',
+						marginTop: '0',
+						paddingLeft: '8px',
+					}}
+					value={storedValue}
+					onChange={(event, val) => {
+						if (val !== null) {
+							searchByCategory(val.id)
+						} else {
+							/* ampSDK.clearValue();
```
Comment (rs-amp) [inline components/ProductSelector/index.tsx:381]: Why is this commented?

File: components/ProductTile/index.tsx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        padding: 6px 10px;
+        transition: all 0.8s ease;
+        border-radius: 3px;
+        background: rgba(255, 255, 255, 0.75);
+        z-index: 1;
+    }
+    &:hover {
+        text-decoration: none;
+        h2 {
+            //text-decoration: underline;
```
Comment (rs-amp) [inline components/ProductTile/index.tsx:73]: unused?

File: components/ProductTile/index.tsx [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+
+    const handleLoaded = () => {
+        setImageLoaded(true)
+    }
+
+    useEffect(() => {
+        const vari = product.selectedVariant ? product.selectedVariant : product.variants[0]
+        //setImageLoaded(false)
```
Comment (rs-amp) [inline components/ProductTile/index.tsx:124]: unused?

File: components/ProductTile/index.tsx [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                        : 
+                        <></>
+                    } */}
+                </div>
+            )    
+        }
+    }
+
+    const edit = () => {
+        //console.log('edit: ', product)
```
Comment (rs-amp) [inline components/ProductTile/index.tsx:170]: unused?

File: components/SortableList/Card.tsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
+import type { Identifier, XYCoord } from 'dnd-core'
+import type { FC } from 'react'
+import { useRef } from 'react'
+import { useDrag, useDrop } from 'react-dnd'
+
+import { ItemTypes } from './ItemTypes'
+import ProductTile from '../ProductTile'
+
+const style = {
+	//border: '1px dashed gray',
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply components/SortableList/Card.tsx:10]: Also this is unused.

File: lib/sdk.ts [hunk 11]
```diff
@@ -12,20 +12,28 @@ const amplienceSDK = async () => {
...
 
     const enforceValue = (value: any) => {
         let val = value
+
         if (typeof value !== 'string') {
             val = value.id
         }
-        return isEnforced() ? `${sdk.field.schema?.pattern.split('/')[0]}/${val}` : val
+        if(isEnforced()) {
+            let pattern = sdk.field.schema?.pattern.split('/')
```
Comment (rs-amp) [inline lib/sdk.ts:25]: Nice catch on this bug.

**Review by rs-amp — COMMENTED** [review]
Looks good, just a few style things and unused commented code around.

data/README.md is empty for some reason.

File: package.json [hunk 12]
```diff
@@ -3,7 +3,7 @@
   "version": "1.1.0",
   "private": true,
   "scripts": {
-    "dev": "next dev",
+    "dev": "next dev --port 3002",
```
[the remark this answers is not in the document]
Comment (sweetenr) [reply package.json:6]: yes

[the remark this answers is not in the document]
Comment (sweetenr) [reply package.json]: reverted

File: components/SortableList/index.tsx [hunk 13]
```diff
@@ -0,0 +1,78 @@
+import React, { useState, useCallback, useEffect } from "react";
+import { ImageList } from "@mui/material";
+import {Card} from "./Card";
+import update from "immutability-helper"
+import {DndProvider} from "react-dnd";
+import {HTML5Backend} from "react-dnd-html5-backend"
+
+export interface Item {
+	id: number
```
[the remark this answers is not in the document]
Comment (sweetenr) [reply components/SortableList/index.tsx:9]: updated

File: components/SortableList/Card.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (sweetenr) [reply components/SortableList/Card.tsx]: updated

File: components/ProductSelector/index.tsx (hunk 6 above)
Comment (sweetenr) [reply components/ProductSelector/index.tsx]: not sure..

File: components/ProductSelector/index.tsx [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+	}
+
+	const searchByKeyword = async () => {
+		setResults([])
+		setLoading(true)
+		const p = await ampSDK.commerceApi.getProducts({
+			keyword: keywordInput.current.value,
+		})
+
+		/* const variants = await ampSDK.commerceApi.getVariants({productId: '25517934M'})
```
[the remark this answers is not in the document]
Comment (sweetenr) [reply components/ProductSelector/index.tsx]: removed

File: components/ProductTile/index.tsx (hunk 7 above)
Comment (sweetenr) [reply components/ProductTile/index.tsx]: removed

File: components/ProductSelector/index.tsx [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+
+	const removeProduct = (productTile: any) => {
+		setSelectedProducts(
+			selectedProducts.filter(
+				(p) => p.deleteKey !== productTile.deleteKey
+			)
+		)
+	}
+
+	//const updateVariant()
```
[the remark this answers is not in the document]
Comment (sweetenr) [reply components/ProductSelector/index.tsx]: update

File: components/ProductSelector/index.tsx (hunk 2 above)
Comment (sweetenr) [reply components/ProductSelector/index.tsx]: I think I tried having that useEffect depend on container and it wasn't working..

**Pull request merged**